File: Classes/GuideStep.m
//
//  GuideStep.m
//  iFixit
//
//  Created by David Patierno on 8/7/10.
//  Copyright 2010 iFixit. All rights reserved.
//

#import "GuideStep.h"


@implementation GuideStep

@synthesize number, title, lines, images, video, embed;

+ (GuideStep *)guideStepWithDictionary:(NSDictionary *)dict {
	GuideStep *guideStep = [[GuideStep alloc] init];
	
	guideStep.number = [dict[@"orderby"] integerValue];
	guideStep.title = dict[@"title"];
	
    // Media
    NSDictionary *media = dict[@"media"];

    // Possible types: image, video, embed
    NSString *type = media[@"type"];

    if ([type isEqual:@"image"]) {
        guideStep.images = [NSMutableArray array];
        NSArray *images = media[@"image"];
        for (NSDictionary *image in images)
            [guideStep.images addObject:[GuideImage guideImageWithDictionary:image]];
    } else if ([type isEqual:@"video"]) {
        NSDictionary *video = media[@"data"];
        guideStep.video = [GuideVideo guideVideoWithDictionary:video];
    } else if ([type isEqual:@"embed"]) {
        NSDictionary *embed = media[@"data"];
        guideStep.embed = [GuideEmbed guideEmbedWithDictionary:embed];
    }

	// Lines
	guideStep.lines = [NSMutableArray array];
	NSArray *lines = [dict valueForKey:@"lines"];
	for (NSDictionary *line in lines)
		[guideStep.lines addObject:[GuideStepLine guideStepLineWithDictionary:line]];
	
	return [guideStep autorelease];
}

- (void)dealloc {
    [title release];
    [lines release];
    [images release];
    [video release];
    [embed release];
    
    [super dealloc];
}

@end

File: Classes/GuideStep.swift
//
//  GuideStep.h
//  iFixit
//
//  Created by David Patierno on 8/7/10.
//  Copyright 2010 iFixit. All rights reserved.
//

import Foundation

class GuideStep: NSObject {

    var number = 0
    var stepid = 0
    var title: String!
    var lines:[GuideStepLine] = []
    var images:[GuideImage] = []
    var video: GuideVideo!
    var embed: GuideEmbed!
    
    init(json: [String:AnyObject]) {
        number = json["orderby"] as! Int
        title = json["title"] as! String
        stepid = json["stepid"] as! Int
        
        // Media
        let media = json["media"] as! [String: AnyObject]
        
        // Possible types: image, video, embed
        let type = media["type"] as? String
        
        switch type! {
            
        case "image":
            for image in media["data"] as! [[String:AnyObject]] {
                images.append(GuideImage(json: image))
            }
            
        case "video":
            let video = media["data"] as! [String:AnyObject]
            self.video = GuideVideo(json: video)
            
        case "embed":
            let embed = media["data"] as! [String:AnyObject]
            self.embed = GuideEmbed(json: embed)
            
        default:
            break
        }
        
        // Lines
        for line in json["lines"] as! [[String:AnyObject]] {
            lines.append(GuideStepLine(json: line))
        }
    }

}
